### PR TITLE
:construction_worker: Added new methods and inline methods

### PR DESCRIPTION
**Pull request by Globegitter**
This adds most of the methods and changes we have been using internally and we think really improves the usability of this library and fixes an imo very important bug we ran into a lot at the beginning in production and is not super easy to debug if you are not aware of it: #59 

One nice thing about having all these inline methods it basically means that we do not need a separate coroutines result but that one can work out of the box with coroutines as you can see in the `.map` test I added.

I know this PR adds a lot at once, I am also happy to split it up and I guess it would require a 4.0 release. Let me know what you think @kittinunf .

### CODE REVIEW

**Comment by kittinunf** [comment]
Wow!!! Thanks for this :clap:. Let me take a look on this.

File: result/src/main/kotlin/com/github/kittinunf/result/Factory.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package com.github.kittinunf.result
+
+inline fun <V> runCatching(block: () -> V): Result<V, Exception> {
```
Comment (kittinunf) [inline result/src/main/kotlin/com/github/kittinunf/result/Factory.kt:3]: Ah this is a good idea 👏

File: result/src/main/kotlin/com/github/kittinunf/result/Factory.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    return try {
+        Result.success(block())
+    } catch (e: Exception) {
+        Result.error(e)
+    }
+}
+
+inline infix fun <T, V> T.runCatching(block: T.() -> V): Result<V, Exception> {
```
Comment (kittinunf) [inline result/src/main/kotlin/com/github/kittinunf/result/Factory.kt:11]: This too 👏

File: result/src/main/kotlin/com/github/kittinunf/result/Result.kt [hunk 3]
```diff
@@ -28,30 +28,57 @@ fun <V: Any?, E: Exception> Result<V, E>.getOrNull(): V? {
...
+}
+
+fun <V : Any?, E : Exception> Result<V, E>.getExceptionOrNull(): E? {
+    return when (this) {
+        is Result.Success -> null
+        is Result.Failure -> error
+    }
+}
+
+inline fun <V : Any?, E : Exception, U : Any?, F : Exception> Result<V, E>.mapEither(success: (V) -> U, failure: (E) -> F): Result<U, F> {
```
Comment (kittinunf) [inline result/src/main/kotlin/com/github/kittinunf/result/Result.kt:45]: In the previous implementation, we use the naming `fold` for this right? Do you think the old naming was not descriptive enough? 🤔

File: result/src/main/kotlin/com/github/kittinunf/result/Result.kt [hunk 4]
```diff
@@ -28,30 +28,57 @@ fun <V: Any?, E: Exception> Result<V, E>.getOrNull(): V? {
     }
 }
 
-inline fun <V : Any?, U : Any?, E : Exception> Result<V, E>.map(transform: (V) -> U): Result<U, E> = try {
+inline infix fun <V : Any?, E : Exception> Result<V, E>.getExceptionOrElse(fallback: (V) -> E): E {
```
Comment (kittinunf) [inline result/src/main/kotlin/com/github/kittinunf/result/Result.kt:31]: Just a question though, in what scenario `getExceptionOrElse` makes sense to use? I personally have never feel that I want to just to get the exception from `Result`?

File: result/src/main/kotlin/com/github/kittinunf/result/Result.kt [hunk 5]
```diff
@@ -64,7 +91,17 @@ inline fun <V : Any?, E : Exception> Result<V, E>.onError(f: (E) -> Unit) = when
     }
 }
 
-fun <V : Any?, E : Exception> Result<V, E>.any(predicate: (V) -> Boolean): Boolean = try {
+inline fun <V : Any?, E : Exception> Result<V, E>.onSuccess(f: (V) -> Unit): Result<V, E> {
```
Comment (kittinunf) [inline result/src/main/kotlin/com/github/kittinunf/result/Result.kt:94]: Do we need to have `onFailure` for the sake of completeness?

File: result/src/main/kotlin/com/github/kittinunf/result/Result.kt (hunk 3 above)
Comment (Globegitter) [reply result/src/main/kotlin/com/github/kittinunf/result/Result.kt:45]: There is a difference, `mapEither` returns another `Result`, so it is basically a shortening for having to chain a `.map` and a `.flatmap`

File: result/src/main/kotlin/com/github/kittinunf/result/Result.kt (hunk 4 above)
Comment (Globegitter) [reply result/src/main/kotlin/com/github/kittinunf/result/Result.kt:31]: Agreed there are not that many cases, but we do have a few, mostly for testing however. Also happy to remove this one if you feel it is not needed enough.

File: result/src/main/kotlin/com/github/kittinunf/result/Result.kt (hunk 5 above)
Comment (Globegitter) [reply result/src/main/kotlin/com/github/kittinunf/result/Result.kt:94]: Yep, there exists already an `onError`.

File: result/src/main/kotlin/com/github/kittinunf/result/Result.kt (hunk 4 above)
Comment (kittinunf) [reply result/src/main/kotlin/com/github/kittinunf/result/Result.kt:31]: Yeah, I think at least I would love to keep the API surface slim. If it is not common enough, do you think it is worthwhile to keep it from the library and have it as a specific extension in client's repo?

File: result/src/main/kotlin/com/github/kittinunf/result/Result.kt (hunk 3 above)
Comment (kittinunf) [reply result/src/main/kotlin/com/github/kittinunf/result/Result.kt:45]: OK!

File: result/src/main/kotlin/com/github/kittinunf/result/Result.kt (hunk 5 above)
Comment (kittinunf) [reply result/src/main/kotlin/com/github/kittinunf/result/Result.kt:94]: Oh sorry I think I missed that 👍

File: result/src/test/kotlin/com/github/kittinunf/result/FactoryTests.kt [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+import org.hamcrest.CoreMatchers
+import org.hamcrest.CoreMatchers.equalTo
+import org.hamcrest.CoreMatchers.instanceOf
+import org.junit.Test
+import org.hamcrest.MatcherAssert.assertThat
+
+class FactoryTests {
+    @Test
+    fun testRunCatching() {
```
Comment (kittinunf) [inline result/src/test/kotlin/com/github/kittinunf/result/FactoryTests.kt:11]: It looks like this test `inline fun <V> runCatching(block: () -> V)` , could you please add another one for another version of `runCatching`?

**Comment by kittinunf** [comment]
@Globegitter Thanks soooo much! I agree with you I think this is a great addition to the library! I think we need 4.0 for this though. but I am happy to bump this up to improve the usage of `Result` 👏

File: result/src/main/kotlin/com/github/kittinunf/result/Result.kt (hunk 4 above)
Comment (Globegitter) [reply result/src/main/kotlin/com/github/kittinunf/result/Result.kt:31]: Yeah I think that is fair enough.

File: result/src/test/kotlin/com/github/kittinunf/result/FactoryTests.kt (hunk 6 above)
Comment (Globegitter) [reply result/src/test/kotlin/com/github/kittinunf/result/FactoryTests.kt:17]: Done.

**Comment by Globegitter** [comment]
Awesome, happy to hear that we can get this in :) And yes makes sense to release as 4.0

**Comment by Globegitter** [comment]
@kittinunf anything still blocking the merge? Would love to upgrade one of our new projects to the new version asap :) I am also not sure why codecov is saying that certain lines are not covered when my tests clearly are covering them. Any idea?

**Comment by kittinunf** [comment]
I don't care too much about the codecov no worries. 

Yes let me check one more time but I think it is pretty much good to go.

**Review by kittinunf — APPROVED** [review]
I think this is golden, let me merge and prepare for 4.0

**Pull request merged**